iothub_client/src/iothub_client_core.c

```diff
@@ -514,69 +526,132 @@ static void iothub_ll_device_twin_callback(DEVICE_TWIN_UPDATE_STATE update_state
+            queue_cb_info.iothub_callback.dev_twin_cb_info.payload = (unsigned char*)malloc(size);
 
-            if (queue_cb_info.iothub_callback.dev_twin_cb_info.payLoad == NULL)
+            if (queue_cb_info.iothub_callback.dev_twin_cb_info.payload == NULL)
             {
                 LogError("Failure allocating payload in get device twin callback.");
                 queue_cb_info.iothub_callback.dev_twin_cb_info.size = 0;
             }
             else
             {
-                (void)memcpy(queue_cb_info.iothub_callback.dev_twin_cb_info.payLoad, payLoad, size);
+                (void)memcpy(queue_cb_info.iothub_callback.dev_twin_cb_info.payload, payload, size);
                 queue_cb_info.iothub_callback.dev_twin_cb_info.size = size;
             }
         }
 
+        // Will copy queue_cb_info into vector.  (Not a vector of pointers, but a vector of USER_CALLBACK_INFOs)
         if (VECTOR_push_back(queue_context->iotHubClientHandle->saved_user_callback_list, &queue_cb_info, 1) != 0)
         {
             LogError("device twin callback userContextCallback vector push failed.");
 
-            if (queue_cb_info.iothub_callback.dev_twin_cb_info.payLoad != NULL)
+            if (queue_cb_info.iothub_callback.dev_twin_cb_info.payload != NULL)
             {
-                free(queue_cb_info.iothub_callback.dev_twin_cb_info.payLoad);
+                free(queue_cb_info.iothub_callback.dev_twin_cb_info.payload);
             }
         }
 
         free(queue_context);
     }
+}
+
+static void iothub_ll_get_device_twin_section_async_callback(DEVICE_TWIN_UPDATE_STATE updateState, IOTHUB_TWIN_RESPONSE_HANDLE twinResponse, const unsigned char* payload, size_t size, void* userContextCallback)
+{
+    IOTHUB_QUEUE_CONSOLIDATED_CONTEXT* queue_context = (IOTHUB_QUEUE_CONSOLIDATED_CONTEXT*)userContextCallback;
+
+    if (queue_context == NULL || twinResponse == NULL)
+    {
+        LogError("Invalid argument (queue_context=%p, twinResponse=%p)", queue_context, twinResponse);
+    }
     else
     {
-        LogError("Get device twin callback userContextCallback NULL");
+        USER_CALLBACK_INFO queue_cb_info;
+        queue_cb_info.type = CALLBACK_TYPE_DEVICE_TWIN;
+        queue_cb_info.userContextCallback = queue_context->userContext;
+        queue_cb_info.iothub_callback.dev_twin_cb_info.updateState = updateState;
+        queue_cb_info.iothub_callback.dev_twin_cb_info.userCallback.getTwinSection = queue_context->userCallback.getTwinSection;
+        queue_cb_info.iothub_callback.dev_twin_cb_info.userContext = queue_context->userContext;
+
+        IOTHUB_TWIN_RESPONSE_HANDLE copyTwinResponse = IoTHubTwin_CreateCopyResponse(twinResponse);
+        if (copyTwinResponse == NULL)
+        {
+            LogError("Failed to create copy of twinResponse");
+            free(queue_context);
+            return;
+        }
+        else
```

Comment:
no need for else?

Reply:
i've updated this and comments, such that if memory allocation fails, the callback is still invoked but NULL is passed for twin_response.  This matches the behavior for the payload.